### DISCUSSION TITLE
Allow environment variables or aliases in preview urls

### DISCUSSION THREAD

**Discussion by samatcd** [craftcms/cms#9975] [Ideas] [answered]
I'm trying to use Craft with Gatsby Cloud, and want to be able to set up development, staging, and production systems. Development will use the local dev server for previews, but both staging and production will use the Gatsby Cloud preview site for its fast incremental builds. This causes an issue in that the preview site has a different URL from the actual website, and there's no way to tell the preview URL what that is without hard-coding them in.

I tried to use an alias to interpolate an environment variable into the preview URL:

`config/general.php`
```
...
    'aliases' => [
        '@previewUrl' => craft\helpers\App::env('PREVIEW_SITE_URL'),
    ],
...
```

And something like this in the relevant project config yaml:

```
previewTargets:
  - __assoc__:
      - - label
        - "Primary entry page"
      - - urlFormat
        - "@previewUrl/{slug}"
      - - refresh
        - ""
```

But the URL won't replace `@previewUrl`. I also tried `"$PREVIEW_SITE_URL/{slug}"` but that had the same effect.

I've got project config changes locked in staging and production so these happen only in development, and Craft is running in headless mode, so I can't just set different preview target domains in each deployed environment — it needs to be done through an environment variable so it can be varied by our CI/CD pipeline.

**Comment by brandonkelly** [comment, marked as answer]
Aliases should work – the URL format is passed through `Craft::parseEnv()` (which parses aliases):

https://github.com/craftcms/cms/blob/f51fff87e02775d370bfd3947e00915a70638760/src/base/Element.php#L2662

`$PREVIEW_SITE_URL/{slug}` would not though, as we don’t support appending additional text to environment variables.

If you go to Utilities → Project Config on production, and search for `"@previewUrl/{slug}"` within “Loaded Project Config Data”, are you seeing that it was in fact applied correctly?
